contributors/design-proposals/storage/svcacct-token-volume-source.md

```diff
@@ -0,0 +1,168 @@
+injected into pods as secrets. This proposal proposes a new mechanism of
+distribution with support for [improved service account
+tokens](https://github.com/kubernetes/community/pull/1460) and explores how to
+migrate from the existing mechanism backwards compatibly.
+
+## Motivation
+
+Many workloads running on Kubernetes need to prove to external parties who they
+are in order to participate in a larger application environment. This identity
+must be attested to by the orchestration system in a way that allows a third
+party to trust that an arbitrary container on the cluster is who it says it is.
+In addition, infrastructure running on top of Kubernetes needs a simple
+mechanism to communicate with the Kubernetes APIs and to provide more complex
+tooling. Finally, a significant set of security challenges are associated with
+storing service account tokens as secrets in Kubernetes and limiting the methods
+whereby malicious parties can get access to these tokens will reduce the risk of
+platform compromise.
+
+As a platform, Kubernetes should evolve to allow identity management systems to
+provide more powerful workload identity without breaking existing use cases, and
+provide a simple out of the box workload identity that is sufficient to cover
+the requirements of bootstrapping low-level infrastructure running on
+Kubernetes. We expect that other systems to cover the more advanced scenarios,
+and see this effort as necessary glue to allow more powerful systems to succeed.
+
+With this feature, we hope to provide a backwards compatible replacement for
+service account tokens that strengthens the security and improves the
+scalability of the platform.
+
+## Proposal
+
+Kubernetes should:
+
+1.  Allow the automatic service account token injection to be disabled via
+    configuration.
+1.  Allow automatic service account token creation to occur only when a user
+    explicitly requests.
+1.  Implement a ServiceAccountToken volume projection that that maintains a
+    service account token requested by the node from the TokenRequest API.
+
+### Token Volume Projection
+
+A new volume projection will be implemented with an API that closely matches the
+TokenRequest API.
+
+```go
+type ProjectedVolumeSource struct {
+  Sources []VolumeProjection
+  DefaultMode *int32
+}
+
+type VolumeProjection struct {
+  Secret *SecretProjection
+  DownwardAPI *DownwardAPIProjection
+  ConfigMap *ConfigMapProjection
+  ServiceAccountToken *ServiceAccountTokenProjection
+}
+
+type ServiceAccountTokenProjection struct {
+  // Audiences are the intendend audiences of the token.
+  Audiences []string
+  // ExpirationSeconds is the requested duration of validity of the service
+  // account token
+  ExpirationSeconds *int64
+  // Path is the relative path of the file to project the token into.
+  Path string
+}
+```
+
+A volume plugin implemented in the kubelet will project a service account token
+sourced from the TokenReview API into volumes created from
```

Comment:
This type is supposed to match TokenRequest. TokenReview here is a typo, thanks for the catch. TokenRequest supports tokens with multiple audiences (for better or worse):

https://github.com/kubernetes/kubernetes/blob/c36e66dd66d04eafeb5e6668b1939244c62720a8/pkg/apis/authentication/types.go#L108-L114

TokenReview doesn't have audiences yet:

https://github.com/kubernetes/kubernetes/blob/c36e66dd66d04eafeb5e6668b1939244c62720a8/pkg/apis/authentication/types.go#L59

 but when it does, it will be an array:

https://github.com/kubernetes/kubernetes/pull/62802

Reply:
cool, thanks for the clarification. I was looking at the design doc [here](https://github.com/mikedanese/community/blob/eb687869434b31ed6fa8c3553d7eea17034133de/contributors/design-proposals/auth/bound-service-account-tokens.md#modify-tokenreviewsauthenticationk8sio) which has a single Audience in the TokenReviewSpec.

Another follow-up question: `Audiences` and `ExpirationSeconds` is filled in by the TokenRequest, right? So the only pre-specified fields are the TokenRequest name (I assume you are using name to identify the TokenRequest) and `Path`?

Reply:
Audiences and ExpirationSeconds are config provided by the pod author, but have defaults of ["k8s-api"] and 1 hour. However, a user might want a token with audience ["vault"] and expiration seconds 3 hours, and they should be able to configure that in the volume projection.

Reply:
Who creates the TokenRequest object? The TokenProjection volume plugin, not the pod author, right?

Reply:
Yes, the token volume plugin (with kubelet's client) makes the TokenRequest.